src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolBlock2.cs

```diff
@@ -98,24 +98,27 @@ protected MemoryPoolBlock2()
         }
 
         /// <summary>
-        /// Called to ensure that a block is pinned, and return the pointer to native memory just after
-        /// the range of "active" bytes. This is where arriving data is read into.
+        /// Called to ensure that a block is pinned, and return the pointer to the native address
+        /// of the first byte of this block's Data memory. Arriving data is read into Pin() + End.
+        /// Outgoing data is read from Pin() + Start.
         /// </summary>
         /// <returns></returns>
         public IntPtr Pin()
         {
-            Debug.Assert(!_pinHandle.IsAllocated);
-
             if (_dataArrayPtr != IntPtr.Zero)
             {
                 // this is a slab managed block - use the native address of the slab which is always locked
-                return _dataArrayPtr + End;
+                return _dataArrayPtr;
+            }
+            else if (_pinHandle.IsAllocated)
```

Comment:
Unpin is still only called once for each block (notice that `Debug.Assert(_pinHandle.IsAllocated);` is still in Unpin).

Pin can called multiple times because it is simpler that finding storing the address externally of blocks which are written multiple times (Only non-overlapping segments of the same block are written of course). The block is only returned once _head moves past the block.